### ISSUE TITLE
Refator and expose wflow_html (original: Using wflow_html with blogdown?)

### ISSUE THREAD

**Issue by docmanny** [workflowr/workflowr#126]
Recently, I've discovered [`blogdown`](https://bookdown.org/yihui/blogdown/), and made a big rewrite of my main site with it. Afterwards, however, I realized that in theory there shouldn't be any reason why one would not be able to use `workflowr` in conjunction with it. Specifically, in the [documentation for blogdown](https://bookdown.org/yihui/blogdown/default-template.html), it talks about how you can provide an alternative template in the YAML header of .Rmd files. 

To test this, I started a `blogdown` project with all the defaults, then initialized a `workflowr` project in it. After setting the publish directory to `docs`, and removing the README.md that `workflowr` adds to the `data` folder, I tried making a new post in `blogdown` with `output: workflowr::wflow_html` in the YAML header (also attempted to include the appropriate settings in the YAML file as per the `workflowr` site's suggestions). While no errors occur, it also doesn't change the output in any way different from a standard blogdown post. Has anyone had any luck with getting the two to get along? 

A "simple" solution would be to create a template or archetype that would format Rmd posts in the `workflowr` style, but such a solution requires a deeper familiarity with the package than I have right now...

**Comment by docmanny** [comment]
Well, the "easy" thing to try was confirm that `blogdown:::renderpage` was at fault, and it was. I forked `blogdown` and made the switch from `'blogdown::html_page'` to `'workflowr::wflow_html'` in the same function you quoted above (see: [here](https://github.com/docmanny/blogdown/blob/master/inst/scripts/render_page.R)). It does in fact fix everything without any issue. I used RStudio Cloud to make a proof-of-concept site and pushed the repo to [docmanny/workflowr-blogdown-experiment](https://github.com/docmanny/workflowr-blogdown-experiment); you can see the resulting site [here](https://docmanny-workflowr-blogdown-experiment.netlify.com). 
Its not a perfect fix, as it builds .Rmd files distinctly from the other files, but its a start! Next is the hard work to actually synthesize the two.

**Comment by docmanny** [comment]
For my own notes: seems like its not a totally perfect solution. It doesn't play nicely with `hugo-academic` and the "dark" theme, for example.

**Comment by docmanny** [comment]
Ok, I've figured out the minimal set of changes needed to get `blogdown` and `workflowr` to play nicely. If we boil down what we (really, I) want from `workflowr` in `blogdown` posts, its the "reproducibility report" and "session info" add-ons that are included without me having to think about it. All that is done by the `pre_knit` function inside of `wflow_html`. I was able to get this to work flawlessly within a minimal `blogdown` site by just adding a few lines to `blogdown` which parses an optional `pre_knit` option the same way that it currently parses a `post_processor` option via text in the `blogdown::html_page` function. At that point, you can add the reproducibility report and the session info by just including the `pre_knit` function definition in the YAML header:
```
output:
  blogdown::html_page:
    pre_knit: |+ 
      function(input, ...) {
      require(workflowr)
      add_bibliography <- function(sessioninfo, lines) {
      stopifnot(is.character(sessioninfo), is.character(lines))
      if (!any(stringr::str_detect(lines, '<div id=[\\\'\\\"]refs[\\\'\\\"]>'))) {
        sessioninfo <- c('', '<div id=\\\"refs\\\"></div>', '', sessioninfo)
      }
      return(sessioninfo)
      }
      frames <- sys.frames()
      e <- frames[[length(frames) - 2]]
      lines_in <- readLines(input)
      tmpfile <- file.path(tempdir(), basename(input))
      e$knit_input <- tmpfile
      wflow_opts <- workflowr:::wflow_options(input)
      header <- rmarkdown::yaml_front_matter(input)
      header_opts <- header$workflowr
      for (opt in names(header_opts)) {
      wflow_opts[[opt]] <- header_opts[[opt]]
      }
      if (!is.null(wflow_opts$knit_root_dir)) {
      if (!R.utils::isAbsolutePath(wflow_opts$knit_root_dir)) {
      wflow_opts$knit_root_dir <- absolute(file.path(dirname(input),
      wflow_opts$knit_root_dir))
      }
      }
      if (is.null(wflow_opts$knit_root_dir)) {
      wflow_opts$knit_root_dir <- dirname(absolute(input))
      }
      if (is.null(e$knit_root_dir)) {
      e$knit_root_dir <- wflow_opts$knit_root_dir
      } else {
      wflow_opts$knit_root_dir <- e$knit_root_dir
      }
      header_delims <- stringr::str_which(lines_in, '^-{3}|^\\\\.{3}')
      if (length(header_delims) >= 2) {
      header_end <- header_delims[2]
      header_lines <- lines_in[seq(header_end)]
      } else {
      header_end <- 0
      header_lines <- NULL
      }
      output_dir <- workflowr:::get_output_dir(directory = dirname(input))
      has_code <- workflowr:::detect_code(input)
      report <- workflowr:::create_report(input, output_dir, has_code, wflow_opts)
      if (has_code && is.numeric(wflow_opts$seed) && length(wflow_opts$seed) == 1) {
      seed_chunk <- c('',
      '```{r seed-set-by-workflowr, echo = FALSE}',
      sprintf('set.seed(%d)', wflow_opts$seed),
      '```',
      '')
      } else {
      seed_chunk <- ''
      }
      if (has_code && wflow_opts$sessioninfo != '') {
      sessioninfo <- c('',
      '## Session information',
      '',
      '```{r session-info-chunk-inserted-by-workflowr}',
      wflow_opts$sessioninfo,
      '```',
      '')
      if (!is.null(header$bibliography)) {
      sessioninfo <- add_bibliography(sessioninfo, lines_in)
      }
      } else {
      sessioninfo <- ''
      }
      lines_out <- c(header_lines,
      '**Last updated:** `r Sys.Date()`',
      report,
      '---',
      seed_chunk,
      lines_in[(header_end + 1):length(lines_in)],
      sessioninfo)
      writeLines(lines_out, tmpfile)
      }
```

A more elegant solution might be to also accept a ready-made function rather than a text definition, and then pass that on via a YAML option. This option would require a small change in `workflowr` making the `pre_knit` and `add_bibliography` functions accessible via `:::`.

**Comment by docmanny** [comment]
The deed is done @jdblischak. See rstudio/blogdown#335 for the inclusion of `pre_knit` as an option in `blogdown::html_page`. `workflowr` now works with `blogdown` if you manually add the function as in my above comment, but it would still be nice to have the `pre_knit` function accessible to external functions in `workflowr`.

**Comment by jdblischak** [comment]
@docmanny Great work! I just did a quick test of your solution using the latest version of blogdown. The reproducibility report and session information indeed get added properly. It appears that the main workflowr features that is missing is the links to past versions of the individual files (since blogdown saves them in a specific location).

One thing I had to fix in your code above was change the call to the internal function `absolute` to `workflowr:::absolute`.

> but it would still be nice to have the pre_knit function accessible to external functions in workflowr.

Agreed! I will investigate refactoring `wflow_html` and exposing the individual functions.

**Comment by jdblischak** [comment]
@docmanny I exported the functions from `wflow_html()` in d8c0b9d. This is currently on the "dev" branch and will soon be included in the next release. I'd appreciate if you could test them out (`wflow_pre_knit()`, `wflow_post_knit()`, `wflow_pre_processor()`). Also, the documentation is minimal since I think you're currently the only one using this. If you get this working smoothly, I'd appreciate documentation PRs to improve the man pages for these functions :-)

**Comment by jdblischak** [comment]
@docmanny This feature is now available in [workflowr 1.2.0](https://github.com/jdblischak/workflowr/releases/tag/v1.2.0). You can install it from the GitHub master branch or wait for it to be built and synced across the CRAN mirrors.
